Reorganize code into modules

IvanIsCoding opened the pull request:
<!--
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [x] I ran rustfmt locally
- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
-->

Closes #300 

The organization steps were:
* Group code by topics in `api.rst`
* For small topics, use `topic.rs`
* For larger topics, create `topic/mod.rs` for the Python implementations and `topic/xyz.rs` for implementations

Details worth sharing are:
* `weight_callable` and `NodeRemoved` stayed at `lib.rs` because they are omnipresent; because they are pure Rust code I thought it was a reasonable place
* `wrap_pyfunction(module_name::function_name)` is currently not supported, so all functions had to be imported
* Some modules from the "Others" section in the docs are small; they could be groupped together but `other_algo` did not feel very descriptive

georgios-ts commented:
Wow, I love this!
Some comments:

- `random_circuit` -> `random_graphs`? I'd also be happy if we merge random graph functions and `generators.rs`.
- Move `simple_path.rs` code (or even `core_number.rs`, `*_adjacency_matrix`) into `connectivity.rs`.
- `topological_sort`, `lexicographical_topological_sort` and `collect_runs` are inside `traversal` folder. I'd move them into `dag_algo`.
- Not a fan of the folder naming: `*_algo`. Can we remove it, i.e `layout_algo` -> `layout`?

IvanIsCoding commented:
> Wow, I love this!
> Some comments:
> 
> * `random_circuit` -> `random_graphs`? I'd also be happy if we merge random graph functions and `generators.rs`.
> * Move `simple_path.rs` code (or even `core_number.rs`, `*_adjacency_matrix`) into `connectivity.rs`.
> * `topological_sort`, `lexicographical_topological_sort` and `collect_runs` are inside `traversal` folder. I'd move them into `dag_algo`.
> * Not a fan of the folder naming: `*_algo`. Can we remove it, i.e `layout_algo` -> `layout`?

I agree `layout_algo` is not the best name. I wanted to name it `layout` but clippy complalins about module inception i.e. `mod layout` inside `layout`. I think I will just disable the warning and go ahead.

I will also try to incorporate the other suggestions

mtreinish commented on src/isomorphism/mod.rs:
```diff
@@ -0,0 +1,327 @@
+// under the License.
+
+#![allow(clippy::float_cmp)]
+#![allow(clippy::module_inception)]
```
To avoid this how about renaming `isomorphism.rs` to `vf2.rs`?

IvanIsCoding replied:
Done

mtreinish commented on src/layout/mod.rs:
```diff
@@ -0,0 +1,569 @@
+// License for the specific language governing permissions and limitations
+// under the License.
+
+#![allow(clippy::module_inception)]
```
What about instead of bundling everything in `layout.rs` if we split up the individual layout methods into separate modules like `bipartite.rs`, `spring.rs`, etc. Then we wouldn't need this clippy allow. Also it would give us a place to put `_spring_layout()` and `_random_layout()` to simplify the mod.rs to just be the pyfunctions.

IvanIsCoding replied:
Done, now there are six files with the layout algorithms and `mod.rs` is mostly just docstrings

mtreinish reviewed (commented):
Overall this is awesome! I'm really happy to get rid of the 5k line (and continuing to grow) lib.rs and put things in real modules. I'd like to move forward on merging this sooner rather than later since it basically conflicts with every other PR. Now that we have a more organized structure we can continue to bikeshed on where things go moving forward, especially as it doesn't actually effect our user API we have the freedom to move things around a lot. I just had 2 inline comments on alternatives to avoid the module inception.

Also, it would be really good if you could add something to:
https://github.com/Qiskit/retworkx/blob/main/CONTRIBUTING.md about the code structure now. Basically about how lib.rs is for the python module exports and global code, where we put pyfunction definitions vs pure rust code, standalone modules vs directories, etc. Once the contributing guide is updated I think this is ready to go.

IvanIsCoding commented:
> Overall this is awesome! I'm really happy to get rid of the 5k line (and continuing to grow) lib.rs and put things in real modules. I'd like to move forward on merging this sooner rather than later since it basically conflicts with every other PR. Now that we have a more organized structure we can continue to bikeshed on where things go moving forward, especially as it doesn't actually effect our user API we have the freedom to move things around a lot. I just had 2 inline comments on alternatives to avoid the module inception.
> 
> Also, it would be really good if you could add something to:
> https://github.com/Qiskit/retworkx/blob/main/CONTRIBUTING.md about the code structure now. Basically about how lib.rs is for the python module exports and global code, where we put pyfunction definitions vs pure rust code, standalone modules vs directories, etc. Once the contributing guide is updated I think this is ready to go.

I incorporated the feedback, now we no longer have module inceptions. 

I also added a small section on `CONTRIBUTING.md`, it is a start on explaining the structure and which file contributors will need to change

mtreinish commented on CONTRIBUTING.md:
````diff
@@ -14,6 +14,74 @@ with Qiskit; the general guidelines and advice still apply here.
+> __NOTE:__  If you create a new `your_module.rs`, remember to declare and import it in `lib.rs`:
+> ```rust
+> mod your_module;
+> use your_module::your_new_function;
````
This was actually correct before with `*`. You need to import everything because the `wrap_pyfunction` macro is looking for the `pyfunction` generated interface function based on the name. But if we don't import everything from `your_module` you'll get an error.

```suggestion
> use your_module::*;
```

mtreinish commented on CONTRIBUTING.md:
```diff
@@ -14,6 +14,74 @@ with Qiskit; the general guidelines and advice still apply here.
+in that file then delegate most of logic by importing and calling pure Rust code from 
+`pure_rust_code.rs` and `more_pure_rust_code.rs`.
+
+> __NOTE:__ Do you still have questions about making your first contribution?
```
I'd probably just open this up to everybody not just first contributors :)

```suggestion
> __NOTE:__ Do you still have questions about making your contribution?
```

mtreinish commented on CONTRIBUTING.md:
```diff
@@ -14,6 +14,74 @@ with Qiskit; the general guidelines and advice still apply here.
 
+### Making changes to the code
+
+Retworkx is implemented primarily in Rust with a thin layer of Python. 
```
```suggestion
Retworkx is implemented primarily in Rust with a thin layer of Python.
```

mtreinish commented on CONTRIBUTING.md:
```diff
@@ -14,6 +14,74 @@ with Qiskit; the general guidelines and advice still apply here.
+### Making changes to the code
+
+Retworkx is implemented primarily in Rust with a thin layer of Python. 
+Because of that, most of your code changes will involve modifications to 
```
```suggestion
Because of that, most of your code changes will involve modifications to
```

mtreinish commented on CONTRIBUTING.md:
```diff
@@ -14,6 +14,74 @@ with Qiskit; the general guidelines and advice still apply here.
+
+#### Adding and changing functions in modules
+
+To add and change functions, you will need to modify module files. Modules contain pyfunctions  
```
```suggestion
To add and change functions, you will need to modify module files. Modules contain pyfunctions
```

mtreinish commented on CONTRIBUTING.md:
```diff
@@ -14,6 +14,74 @@ with Qiskit; the general guidelines and advice still apply here.
+#### Adding and changing functions in modules
+
+To add and change functions, you will need to modify module files. Modules contain pyfunctions  
+that will be exported, and can be defined either as a single file such as `tiny.rs` or as a 
```
```suggestion
that will be exported, and can be defined either as a single file such as `tiny.rs` or as a
```

mtreinish commented on CONTRIBUTING.md:
```diff
@@ -14,6 +14,74 @@ with Qiskit; the general guidelines and advice still apply here.
+Rust code. To change an existing function, search for its name and edit the code that
+already exists.
+
+If you want to add a new function, find the module you'd like to insert it in 
```
```suggestion
If you want to add a new function, find the module you'd like to insert it in
```

mtreinish commented on CONTRIBUTING.md:
```diff
@@ -14,6 +14,74 @@ with Qiskit; the general guidelines and advice still apply here.
+and splitting them following the structure of `large/`.
+
+Module directories have a `mod.rs` file containing the pyfunctions. The pyfunctions
+in that file then delegate most of logic by importing and calling pure Rust code from 
```
```suggestion
in that file then delegate most of logic by importing and calling pure Rust code from
```

mtreinish reviewed (approved):
Overall this is great. Thanks for making those updates quickly. I've got a few inline nits on the contributor docs but nothing major, I'll just apply them and merge this! Then we have the fun of rebasing everything :)

mtreinish commented:
I was just curious if there was an impact on compile time, I just did side by side testing locally and there was no difference for me with this PR applied.
